### PR TITLE
Abstraction of threading primitives API 

### PR DESCRIPTION
**Pull request by redtangent**
## Description

This PR makes several changes to improve the platform abstraction of the threading primitives:

 * allows threading functions to be defined at link time so it's not necessary to call the function `mbedtls_threading_set_alt()` to set them
 * tidies up some comments in the threading abstraction layer, removing obsolete checks in `check_config.h` 
 * changes the model of how threading primitive functions are accessed, thereby they are statically linked with `MBEDTLS_THREADING_PTHREAD` and dynamically determined for `MBEDTLS_THREADING_ALT`. This can now be extended for other implementations of the threading for other platforms.

## Status
**READY**

## Requires Backporting
 NO  

This is a functional enhancement to the API and is not appropriate to backport.

## Migrations
YES

## Additional comments
This PR replaces #697, and takes into all review feedback on that PR.

## Todos
- [ n/a] Tests
- [ n/a] Documentation
- [ DONE] Changelog updated
- [ n/a] Backported

### CODE REVIEW

**Comment by sbutcher-arm** [comment]
@yanesca and @andresag01 - picking up where I left off. Can you please review?

Hi @k-stachowiak, it was my intention that we'll use these changes to integrate mbed OS threading into Mbed TLS, and I anticipate you'll be doing that work, so could you also please review this?

File: library/threading.c [hunk 1]
```diff
@@ -69,69 +69,115 @@ static int threading_mutex_unlock_pthread( mbedtls_threading_mutex_t *mutex )
...
 
 /*
  * Define global mutexes
  */
-#ifndef MUTEX_INIT
-#define MUTEX_INIT
-#endif
-mbedtls_threading_mutex_t mbedtls_threading_readdir_mutex MUTEX_INIT;
-mbedtls_threading_mutex_t mbedtls_threading_gmtime_mutex MUTEX_INIT;
+#ifdef MBEDTLS_MUTEX_INITIALIZER
```
Comment (andresag01) [inline library/threading.c]: I suggest changing this to fit the mbedTLS style of using `#if defined(...)` instead...

File: library/threading.c [hunk 2]
```diff
@@ -30,15 +30,15 @@
 #include "mbedtls/threading.h"
 
 #if defined(MBEDTLS_THREADING_PTHREAD)
-static void threading_mutex_init_pthread( mbedtls_threading_mutex_t *mutex )
+void mbedtls_mutex_init( mbedtls_threading_mutex_t *mutex )
 {
     if( mutex == NULL )
```
Comment (k-stachowiak) [inline library/threading.c]: I'm not certain if these tests are necessary. Adding them not only introduces branches but also enforces virtually repeating the same test in the client code as for each of the calls to this API the client must check for the error code. It may be better to put the responsibility of providing valid mutexes on the client an only check for the validity once instead of every time these functions are called.

File: include/mbedtls/threading.h [hunk 3]
```diff
@@ -35,22 +35,129 @@
...
 #endif
 
-#define MBEDTLS_ERR_THREADING_FEATURE_UNAVAILABLE         -0x001A  /**< The selected feature is not available. */
-#define MBEDTLS_ERR_THREADING_BAD_INPUT_DATA              -0x001C  /**< Bad input parameters to function. */
-#define MBEDTLS_ERR_THREADING_MUTEX_ERROR                 -0x001E  /**< Locking / unlocking / free failed with error code. */
+#if defined(MBEDTLS_THREADING_C)
+
+#define MBEDTLS_ERR_THREADING_FEATURE_UNAVAILABLE   0x001A
+                            /**< The selected feature is not available. */
+#define MBEDTLS_ERR_THREADING_BAD_INPUT_DATA        0x001C
```
[the remark this answers is not in the document]
Comment (redtangent) [reply include/mbedtls/threading.h]: Oops. That's embarassing. I must have accidentally removed the signs while changing the formatting. Now restored.

File: include/mbedtls/threading.h [hunk 4]
```diff
@@ -35,22 +35,129 @@
...
+
+#else
+
+/* For no given implementation, define as an opaque pointer */
+typedef struct _mbedtls_threading_mutex_t
+{
+    void* mutex;
+} mbedtls_threading_mutex_t;
+
 #endif
```
[the remark this answers is not in the document]
Comment (redtangent) [reply include/mbedtls/threading.h]: You missed the `#else`. It's actually `!MBEDTLS_THREADING_ALT && !MBEDTLS_THREADING_PTHREAD`. 

Added.

File: library/threading.c [hunk 5]
```diff
@@ -69,69 +69,115 @@ static int threading_mutex_unlock_pthread( mbedtls_threading_mutex_t *mutex )
...
 }
 
-void (*mbedtls_mutex_init)( mbedtls_threading_mutex_t * ) = threading_mutex_init_pthread;
-void (*mbedtls_mutex_free)( mbedtls_threading_mutex_t * ) = threading_mutex_free_pthread;
-int (*mbedtls_mutex_lock)( mbedtls_threading_mutex_t * ) = threading_mutex_lock_pthread;
-int (*mbedtls_mutex_unlock)( mbedtls_threading_mutex_t * ) = threading_mutex_unlock_pthread;
+#define MBEDTLS_MUTEX_INIT_MACRO       threading_mutex_init_pthread
+#define MBEDTLS_MUTEX_FREE_MACRO       threading_mutex_free_pthread
+#define MBEDTLS_MUTEX_LOCK_MACRO       threading_mutex_lock_pthread
+#define MBEDTLS_MUTEX_UNLOCK_MACRO     threading_mutex_unlock_pthread
```
[the remark this answers is not in the document]
Comment (redtangent) [reply library/threading.c]: This is redundant code and has been removed.

If you'd missed it, I changed the functions to be statically linked when `MBEDTLS_THREADING_PTHREAD` is defined rather than using the function pointers which are now only used when `MBEDTLS_THREADING_ALT` is defined. 

These lines have now been removed.

File: library/threading.c [hunk 6]
```diff
@@ -69,69 +69,115 @@ static int threading_mutex_unlock_pthread( mbedtls_threading_mutex_t *mutex )
...
+
+#else
+
+#if defined(MBEDTLS_FS_IO)
+mbedtls_threading_mutex_t mbedtls_threading_readdir_mutex;
+#endif /* MBEDTLS_FS_IO */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE)
+mbedtls_threading_mutex_t mbedtls_threading_gmtime_mutex;
+#endif /* MBEDTLS_HAVE_TIME_DATE */
```
[the remark this answers is not in the document]
Comment (redtangent) [reply library/threading.c]: If you look at the lines immediately preceding these, you'll see that if `MBEDTLS_MUTEX_INITIALIZER` is not defined than you don't initialise them statically, and if it is defined you do. The choice is down to whomever is providing the alternative implementation of the threading primitives.

`mbedtls_mutex_init()` is called extensively through the library whenever the context a mutex is held in, is initialised.

There are two global mutexes, initialised in `threading.c` statically. If they need to be initialised they'll need to be initialised with `MBEDTLS_MUTEX_INITIALIZER` or outside the library. 

Or we could initialise them in the platform initialisation?

**Comment by redtangent** [comment]
Hi @andresag01,

To respond to some of your review comments:

> I think this might not compile when `MBEDTLS_THREADING_PTHREAD` is set because the `_MACRO`macros are still pointing to the old name of the pthread functions.

Yes, it does compile. That was redundant code that should have been removed.

> I think this is a preexisting problem, but I find the threading configuration macros very confusing (excluding the `MBEDTLS_THREADING_C`).

To explain what each of the macros does (which I have slightly tweaked).

 * `MBEDTLS_THREADING_C` - enables the library to use mutexes so it can be used by multi-threaded application code, and so common resources between threads are protected. Without this, Mbed TLS is only single threaded.
 * `MBEDTLS_THREADING_PTHREAD` - enables the threading primitives (mutexes) to be implemented using pthreads.
 * `MBEDTLS_THREADING_ALT` - enables the application code to dynamically define it's own threading primitives using the function `mbedtls_threading_set_alt()`.

So why allow `MBEDTLS_THREADING_C` to be defined, but not `MBEDTLS_THREADING_PTHREAD` or `MBEDTLS_THREADING_ALT`? You will have unresolved dependencies for the threading primitives defined in `threading.h`, but *that's okay* because this is a library and users can provide their own implementations outside of our source code and our build system.

> I think it would be good to add tests to this PR to test the actual ALT interface and not necessarily the concurrency aspect (at this stage). 

Currently the library has no tests for `_alt.h` implementations. I tried creating one, but the library can't build unless it can find the `_alt.h` header, which needs changes to the build files.

I think the project needs to decide how such tests can be built and change the Makefiles accordingly before I can provide a test, otherwise you'll end up with a hack to the include files to find the header in `tests/suites`. 

> I have my strong reservations about the MBEDTLS_MUTEX_INITIALIZER which is not available in some platforms

The project currently supports pthreads and the `_ALT` implementation. The global mutexes for that will need to be initialised by platform code. I think this is an issue for the project documentation rather than this PR.

File: library/threading.c (hunk 2 above)
Comment (sbutcher-arm) [reply library/threading.c]: Do you mean the parameter validation checks? eg. 
```
if( mutex == NULL )
{ 
...
}
```

We've had strong feedback from some users that their absence on many functions is poor practice, and they want to see a lot more. Conversely, some users see them as redundant and unnecessary and taking up precious code space. (And there's another camp that thinks they should be `asserts`).

I think the code can stay as it is for the moment (eg. no change), but at some point we'll need to review the library and introduce more checks but make them optional at compile time (similar to asserts).

**Comment by sbutcher-arm** [comment]
And to add more - I see these changes as laying the groundwork for native threading support in Windows and Mbed OS, for which Windows has an outstanding PR.

**Comment by redtangent** [comment]
As an experiment (or proof of concept), I just added a test suite for threading which you can see on a branch [here](https://github.com/redtangent/mbedtls/tree/threading-model-tests) or the diffs [here](https://github.com/ARMmbed/mbedtls/compare/development...redtangent:threading-model-tests).

Note the introduction of a `TEST` flag to the makefiles to include the test header `threading_alt.h`.   Essentially, by building with an `_alt.h` test header, you end up with a test library, that's good for testing the library, but the library you get is not really usable for anything else. That's a change from the current system.

Anyway, offered up for feedback to discuss how it might work.

File: library/threading.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresag01) [reply library/threading.c]: Ok, if that is our policy it would be good to document that somewhere visible as you commented below. I looked around in the threading files, config.h and knowledgebase article and could not find anything that warns the user/porter that if there is no static initialiser then the user is expected to initialise manually from outside the library or the platform initialisation code.

[the remark this answers is not in the document]
Comment (mpg) [reply library/threading.c]: IMO it's bad to require users to manually initialize global mutexes, because that means they need to change their code whenever we change the set of global mutexes. Currently they're initialized in `mbedtls_threading_set_alt()`. If there are use case where `MBEDTLS_MUTEX_INITIALIZER` is not defined and this function is not called, then I suggest we split its last two lines into a new public function and require users in that situation to call that function, instead of directly accessing global variables.

**Comment by sbutcher-arm** [comment]
@yanesca, @andresag01, @k-stachowiak - can you all please re-review and approve or reject? Thanks

**Comment by k-stachowiak** [comment]
@redtangent Thanks for providing the example. It helps with understanding the possible use cases. I didn't find issues in the code, but I have some remarks about the things I found confusing:

1. In the test_suite_threading.function file there is the pthread.h header compiled in conditionally. Is that always necessary in the client code? Shouldn't the client only set the MBEDTLS_THREADING_PTHREAD symbol and leave the header including to mbedtls?

2. In the lock_mutex case the MBEDTLS_MUTEX_INITIALIZER is used. I am surprised to see this symbol in such context. Is this one of the intended uses for the macro? I thought that only two things happen to this symbol:
- client optionally defines it
- if it's defined then the library initializes the global mutexes statically
I can also imagine client using this symbol for their convenience. Was that the intent?

To summarize I think that the documentation (I assume it's the header comments) would benefit from adding specific instructions about the two degrees of freedom that are offered:
- the selection of the threading implementation,
- the use of the static initializers.

File: library/threading.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply library/threading.c]: It looks like the main incentive to use the macros is to avoid calling `mbedtls_threading_set_alt()`. So we might as well require an all or nothing approach. (We could enforce in check-config.h that either all of these macros are defined or none of them)

**Comment by tom-daubney-arm** [comment]
We are opting to close this PR. We haven't heard back from the author and a lot of time has passed so there is significant bit rot. Additionally with build failures, this PR will require significant rework. Thanks to the author for their time and effort in creating this PR.

**Comment by simonbutcher** [comment]
> We haven't heard back from the author 

This is an Arm PR that has never progressed because it never merited resource from Arm. Saying you never heard back from the author sounds a bit disingenuous, although I'm sure this is just a misunderstanding.

**Comment by daverodgman** [comment]
Apologies @simonbutcher, I don't think your redtangent account was well-known within the team - we will keep an eye out in future reviews of historical PRs.
